firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalSerializer.java

```diff
@@ -271,12 +271,25 @@ TargetData decodeTargetData(com.google.firebase.firestore.proto.Target targetPro
   }
 
   public com.google.firestore.proto.BundledQuery encodeQuery(Query query) {
-    // We store LimitToLast as metadata information below. We remove it from the query itself as
-    // `toTarget()` flips the order by constraints otherwise.
-    Query queryWithoutLimitToLast =
-        query.hasLimitToLast() ? query.limitToFirst(query.getLimitToLast()) : query;
+    // We don't use query.toTarget() here as we want to save the original query, and not the flipped
+    // query that we may send to the backend (if limitToLast() is used). This allows us to
+    // deserialize named queries again without inverting the order by constraints on read.
+    long limit =
```

Comment:
This is better!

What do you think about eliminate `Query` in the picture completely and use `Target` only?

I am thinking:

- Make `NamedQuery` hold `Target` and a limit type enum.
- `BundleSerializer.decodeQuery` becomes `decodeTarget`.
-  This method here becomes `encodeNamedQuery`, the `Target` and limitType should both be available. The `decodeQuery` below becomes `decodeBundledQuery`.

Reply:
Above should also make porting the JS test of saving and reading a limit to last query possible?

Reply:
NamedQuery now contains a BundledQuery model. This matches the Protobuf and gets rid of all this indirection. I didn't go with Target/LimitType in the NameQuery as I needed a return type for the methods in LocalSerializer, which decodes both the limit type and the name query.

There are two tests that test the round-trip encoding. I will go through your "bundle to master" merge PR when I am close to finishing it up to ensure that I am not missing tests.